bot/utils/redis_cache.py

```diff
@@ -120,41 +130,49 @@ async def get(self, key: ValidRedisType, default: Optional[ValidRedisType] = Non
         if value is None:
             return default
         else:
-            value = self._from_typestring(value.decode("utf-8"))
+            value = self._from_typestring(value)
             return value
 
     async def delete(self, key: ValidRedisType) -> None:
         """Delete an item from the Redis cache."""
-        # await self._redis.hdel(self._namespace, key)
+        await self._validate_cache()
+        key = self._to_typestring(key)
+        return await self._redis.hdel(self._namespace, key)
 
     async def contains(self, key: ValidRedisType) -> bool:
         """Check if a key exists in the Redis cache."""
-        # return await self._redis.hexists(self._namespace, key)
+        await self._validate_cache()
+        key = self._to_typestring(key)
+        return await self._redis.hexists(self._namespace, key)
 
     async def items(self) -> AsyncIterator:
         """Iterate all the items in the Redis cache."""
-        # data = await redis.hgetall(self.get_with_namespace(key))
-        # for item in data:
-        #     yield item
+        await self._validate_cache()
+        data = await self._redis.hgetall(self._namespace)  # Get all the keys
+        for key, value in self._dict_from_typestring(data).items():
+            yield key, value
```

Comment:
Neat. Looks beautiful.

Reply:
Yeah I really love this syntax for asynciterators that @jos-b found, it's very elegant.

Reply:
The only reason I see why an async iterator is used is so we can do `async for x in y()` rather than `for x in await y()`, right? It feels misleading because, to me, an async iterator implies it's lazy, which of course `hgetall` is not.

Reply:
Yep, this has now been refactored into returning an ItemsView, @MarkKoz